Fix missing active state for outline buttons

theodorejb opened the pull request:
### Description

Previously there was no visual feedback when clicking an outline button, which made them feel somewhat buggy/broken. This pull request adds an active state to outline buttons so they respond similarly to other buttons when selected.

It also updates the hover state for outline buttons to match that of regular buttons. Previously if an outline button was next to a regular button, when hovered it looked the same as the non-hovered regular button which could cause confusion.

### Type of changes

- [x] Bug fix (non-breaking change which fixes an issue)

### Checklist

- [x] I have read the [contributing guidelines](https://github.com/twbs/bootstrap/blob/main/.github/CONTRIBUTING.md)
- [x] My code follows the code style of the project _(using `npm run lint`)_
- [ ] My change introduces changes to the documentation
- [ ] I have added tests to cover my changes
- [x] All existing tests passed

#### Live previews

- <https://deploy-preview-39099--twbs-bootstrap.netlify.app/docs/5.3/components/buttons/#outline-buttons>

### Related issues

Resolves #39085

theodorejb committed:
Fix missing active state for outline buttons

Resolves #39085

xeron commented:
Please test with the following:

```html
<button type="button" class="btn btn-outline-light active" data-bs-toggle="button" aria-pressed="true">Light</button>
```

I tried your fix and it doesn't look correct – active button state is shaded which doesn't match normal button where hover is shaded. I think hover supposed to be shaded like in usual button, not active-background?

xeron commented:
NVM I just realized normal button active state is shaded too 🤦

xeron commented:
Another issue to think about is mobile devices behavior – this fix makes inactive hovered button brighter then active which helps a bit, but it's still confusing to the end user as they're not sure if state actually changed or not because button doesn't look like inactive one (because it's stays hovered until user taps something else).

theodorejb commented:
@xeron I'm not sure what you you're referring to as confusing; I tested on mobile and it seems to work as expected. This patch doesn't affect hover state - it only adds the missing active state (matching the active state of regular buttons).

xeron commented:
What confusing on mobile is after clicking the button it stays highlighted because of the sticky hover effect:

https://codepen.io/iXeron/pen/xxmGazY

![Screenshot 2023-08-23 at 4 42 23 PM](https://github.com/twbs/bootstrap/assets/110569/86676904-f8b0-4a96-8a59-68a049694039)

https://css-tricks.com/solving-sticky-hover-states-with-media-hover-hover/

theodorejb commented:
> What's confusing on mobile is after clicking the button it stays highlighted because of the sticky hover effect

I see. Note that regular buttons have this sticky-hover behavior on mobile too, so it's not something specific to outline buttons (nor is it affected by this pull request).

My goal with this patch is to specifically address the missing active state on outline buttons, and not change anything else. As you said the sticky-hover effect may warrant a separate design discussion.

theodorejb commented:
@mdo Is there any chance this fix could be included in 5.3.2?

CC @julien-deramond

julien-deramond commented:
@theodorejb The content of the v5.3.2 is now fixed in order to release the patch regarding Dart Sass really soon. The next version will prolly be v5.3.3. I'll integrate the review of this PR for this v5.3.3.

theodorejb committed:
Fix outline button hover color to match regular buttons

Previously the hover state was the same as the non-hover state for regular buttons, which could cause confusion when a hovered outline button is near a regular button.

theodorejb commented:
I added a second commit which updates the outline button hover color to match that of regular buttons. Previously the hover state was the same as the non-hover state of regular buttons, which could cause confusion when hovering an outline button placed next to a non-outline button. This also improves the active state appearance since it no longer looks too dark in comparison to the hover color.

mdo commented:
This incorrectly changes the hover color of outline buttons, so we can't merge that at the moment.

theodorejb commented:
@mdo Should I split the second commit which changes the hover color into a separate pull request? The issue without it is that there's no visual difference between a hovered outline button and a non-hovered regular button, which can be confusing.

julien-deramond commented:
Deferring the review of this PR to v5.4.0, as changing the interface of the `button-outline-variant()` mixin isn't suitable for a patch release.

mdo commented:
v6's .btn-outline already renders distinct hover and active states via oklch (see #39085), so this active-state fix is already covered. Closing as resolved by the v6 rewrite.
